flink-core/src/main/java/org/apache/flink/util/FileUtils.java

```diff
@@ -107,6 +120,89 @@ public static void writeFileUtf8(File file, String contents) throws IOException
 		writeFile(file, contents, "UTF-8");
 	}
 
+	/**
+	 * Reads all the bytes from a file. The method ensures that the file is
+	 * closed when all bytes have been read or an I/O error, or other runtime
+	 * exception, is thrown.
+	 *
+	 * <p>This is an implementation that follow {@link java.nio.file.Files#readAllBytes(java.nio.file.Path)},
+	 * and the difference is that it limits the size of the direct buffer to avoid
+	 * direct-buffer OutOfMemoryError. When {@link java.nio.file.Files#readAllBytes(java.nio.file.Path)}
+	 * or other interfaces in java API can do this in the future, we should remove it.
+	 *
+	 * @param path
+	 *        the path to the file
+	 * @return a byte array containing the bytes read from the file
+	 *
+	 * @throws IOException
+	 *         if an I/O error occurs reading from the stream
+	 * @throws OutOfMemoryError
+	 *         if an array of the required size cannot be allocated, for
+	 *         example the file is larger that {@code 2GB}
+	 */
+	public static byte[] readAllBytes(java.nio.file.Path path) throws IOException {
+		try (SeekableByteChannel channel = Files.newByteChannel(path);
+			InputStream in = Channels.newInputStream(channel)) {
+
+			long size = channel.size();
+			if (size > (long) MAX_BUFFER_SIZE) {
+				throw new OutOfMemoryError("Required array size too large");
+			}
+
+			return read(in, (int) size);
+		}
+	}
+
+	/**
+	 * Reads all the bytes from an input stream. Uses {@code initialSize} as a hint
+	 * about how many bytes the stream will have and uses {@code directBufferSize}
+	 * to limit the size of the direct buffer used to read.
+	 *
+	 * @param source
+	 *        the input stream to read from
+	 * @param initialSize
+	 *        the initial size of the byte array to allocate
+	 * @return a byte array containing the bytes read from the file
+	 *
+	 * @throws IOException
+	 *         if an I/O error occurs reading from the stream
+	 * @throws OutOfMemoryError
+	 *         if an array of the required size cannot be allocated
+	 */
+	private static byte[] read(InputStream source, int initialSize) throws IOException {
+		int capacity = initialSize;
+		byte[] buf = new byte[capacity];
```

Comment:
don't see how we're solving the issue. capacity is still the size of the entire file.

The only difference is that we pass `BUFFER_SIZE` to `InputStream#read`, but that method never allocates an array.

Reply:
You're still talking about OutOfMemoryError on the **heap memory**. This PR is to fix OutOfMemoryError on the **direct memory** . Please take a closer look at the exception stack, and the exception is caused that an overly large size **(capacity - nread)** is passed to the read() method.

_Caused by: java.lang.OutOfMemoryError: **Direct buffer memory**
  at **java.nio.Bits.reserveMemory**(Bits.java:706)
  at java.nio.DirectByteBuffer.<init>(DirectByteBuffer.java:123)
  at java.nio.ByteBuffer.allocateDirect(ByteBuffer.java:311)
  at sun.nio.ch.Util.getTemporaryDirectBuffer(Util.java:241)
  at sun.nio.ch.IOUtil.read(IOUtil.java:195)
  at sun.nio.ch.FileChannelImpl.read(FileChannelImpl.java:182)
  at sun.nio.ch.ChannelInputStream.read(ChannelInputStream.java:65)
  at sun.nio.ch.ChannelInputStream.read(ChannelInputStream.java:109)
  at sun.nio.ch.ChannelInputStream.read(ChannelInputStream.java:103)
  at java.nio.file.Files.read(Files.java:3105)
  at java.nio.file.Files.readAllBytes(Files.java:3158)
  at org.apache.flink.runtime.deployment.TaskDeploymentDescriptor.loadBigData(TaskDeploymentDescriptor.java:338)_

Reply:
Yes, it is passed to `InputStream#read(byte[], int, int)`, but the length is not passed on further to `InputStream#read()`. `len` is only used to check how often we should call `InputStream#read()` in one call to `InputStream#read(byte[], int, int)`, nothing else.

All you're doing is increase calls to `InputStream#read(byte[], int, int)`; this however has no functional effect.

Reply:
hold on

Reply:
Please look at the exception stack,  here is ChannelInputStream instance, so you should check the code of ChannelInputStream.